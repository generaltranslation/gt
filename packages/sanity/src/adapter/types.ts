export type IgnoreFields = {
  documentId?: string;
  fields?: { property: string; type?: string }[];
};

export type TranslateDocumentFilter = {
  documentId?: string;
  type?: string;
};
