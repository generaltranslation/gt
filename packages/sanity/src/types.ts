// adapted from https://github.com/sanity-io/sanity-translations-tab. See LICENSE.md for more details.

import { SanityClient, Schema, TypedObject } from 'sanity';
import type { SerializedDocument } from './serialization';
import { PortableTextTypeComponent } from '@portabletext/to-html';
import { DeserializerRule } from '@sanity/block-tools';

export type TranslationTaskLocaleStatus = {
  localeId: string;
  progress: number;
};

export type TranslationTask = {
  document: GTFile;
  locales: TranslationTaskLocaleStatus[];
  linkToVendorTask?: string;
};

export type TranslationLocale = {
  localeId: string;
  description: string;
  enabled?: boolean;
};

//this varies according to provider
//not every vendor uses every field
export type Secrets = {
  organization: string;
  project: string;
  token?: string;
  secret?: string;
  username?: string;
  password?: string;
  proxy?: string;
};

export type WorkflowIdentifiers = {
  workflowUid: string;
  workflowName: string;
};

export type GTFile = {
  documentId: string;
  versionId?: string;
};

export interface Adapter {
  getLocales: (secrets: Secrets | null) => Promise<TranslationLocale[]>;
  getTranslationTask: (
    document: GTFile,
    secrets: Secrets | null
  ) => Promise<TranslationTask>;
  createTask: (
    documentInfo: GTFile,
    serializedDocument: GTSerializedDocument,
    localeIds: string[],
    secrets: Secrets | null,
    workflowUid?: string,
    callbackUrl?: string
  ) => Promise<TranslationTask>;
  getTranslation: (
    document: GTFile,
    localeId: string,
    secrets: Secrets | null
  ) => Promise<any | null>;
}

export interface TranslationFunctionContext {
  client: SanityClient;
  schema: Schema;
}

export type GTSerializedDocument = Omit<SerializedDocument, 'name'> & GTFile;

export type ExportForTranslation = (
  documentInfo: GTFile,
  context: TranslationFunctionContext
) => Promise<GTSerializedDocument>;

export type ImportTranslation = (
  documentInfo: GTFile,
  localeId: string,
  document: string,
  context: TranslationFunctionContext,
  mergeWithTargetLocale?: boolean,
  publish?: boolean
) => Promise<void>;

export type TranslationsTabConfigOptions = {
  adapter: Adapter;
  secretsNamespace: string | null;
  exportForTranslation: ExportForTranslation;
  importTranslation: ImportTranslation;
  serializationOptions?: {
    additionalStopTypes?: string[];
    additionalSerializers?: Record<
      string,
      PortableTextTypeComponent | undefined
    >;
    additionalDeserializers?: Record<
      string,
      (value: HTMLElement) => TypedObject
    >;
    additionalBlockDeserializers?: DeserializerRule[];
  };
  workflowOptions?: WorkflowIdentifiers[];
  localeIdAdapter?: (id: string) => string;
  languageField?: string;
  callbackUrl?: string;
  mergeWithTargetLocale?: boolean;
};
