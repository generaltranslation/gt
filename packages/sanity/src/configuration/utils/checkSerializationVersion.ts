// adapted from https://github.com/sanity-io/sanity-translations-tab. See LICENSE.md for more details.

export const checkSerializationVersion = (HTMLdoc: string): string | null => {
  const parser = new DOMParser();
  const node = parser.parseFromString(HTMLdoc, 'text/html');
  const versionMetaTag = Array.from(node.head.children).find(
    (metaTag) => metaTag.getAttribute('name') === 'version'
  );
  if (!versionMetaTag) {
    return null;
  }

  const version = versionMetaTag.getAttribute('content');
  return version;
};
