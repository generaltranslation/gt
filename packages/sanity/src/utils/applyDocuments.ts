import { JSONPath } from 'jsonpath-plus';
import JSONPointer from 'jsonpointer';
import type { IgnoreFields } from '../adapter/types';

export function applyDocuments(
  documentId: string,
  sourceDocument: Record<string, any>,
  targetDocument: Record<string, any>,
  ignore: IgnoreFields[]
) {
  const ignoreFields = ignore.filter(
    (field) =>
      field.documentId === documentId ||
      field.documentId === undefined ||
      field.documentId === null
  );

  // Start with a shallow copy of the source document
  const mergedDocument = { ...sourceDocument };

  // Merge top-level properties of targetDocument
  for (const [key, value] of Object.entries(targetDocument)) {
    mergedDocument[key] = value;
  }

  // Process ignored fields and restore them from source document
  for (const ignoreField of ignoreFields) {
    if (!ignoreField.fields) continue;

    for (const field of ignoreField.fields) {
      const { property, type } = field;

      try {
        // Use JSONPath to find matching paths, then JSONPointer to get/set values
        const sourceResults = JSONPath({
          json: sourceDocument,
          path: property,
          resultType: 'all',
          flatten: true,
          wrap: true,
        });

        if (sourceResults && sourceResults.length > 0) {
          // Process each matching path
          sourceResults.forEach((result: { pointer: string; value: any }) => {
            const sourceValue = result.value;

            // If type is specified, check if it matches the object's _type property
            if (type !== undefined) {
              if (
                typeof sourceValue === 'object' &&
                sourceValue !== null &&
                sourceValue._type === type
              ) {
                // Type matches, restore the entire object using JSONPointer
                JSONPointer.set(mergedDocument, result.pointer, sourceValue);
              }
            } else {
              // No type specified, restore the value using JSONPointer
              JSONPointer.set(mergedDocument, result.pointer, sourceValue);
            }
          });
        }
      } catch (error) {
        // Invalid JSONPath, skip this field
        console.warn(`Invalid JSONPath: ${property}`, error);
      }
    }
  }

  return mergedDocument;
}
