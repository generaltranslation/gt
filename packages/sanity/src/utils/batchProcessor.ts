import { GTFile, TranslationFunctionContext } from '../types';
import { importDocument } from '../translation/importDocument';

export interface BatchProcessorOptions {
  batchSize?: number;
  onProgress?: (current: number, total: number) => void;
  onItemSuccess?: (item: any, result: any) => void;
  onItemFailure?: (item: any, error: any) => void;
}

export interface ImportBatchItem {
  docInfo: GTFile;
  locale: string;
  data: any;
  translationContext: TranslationFunctionContext;
  key: string;
}

export async function processBatch<T>(
  items: T[],
  processor: (item: T) => Promise<any>,
  options: BatchProcessorOptions = {}
): Promise<{
  successCount: number;
  failureCount: number;
  successfulItems: any[];
  failedItems: { item: T; error: any }[];
}> {
  const { batchSize = 20, onProgress, onItemSuccess, onItemFailure } = options;

  let successCount = 0;
  let failureCount = 0;
  const successfulItems: any[] = [];
  const failedItems: { item: T; error: any }[] = [];

  for (let i = 0; i < items.length; i += batchSize) {
    const batch = items.slice(i, i + batchSize);

    const batchPromises = batch.map(async (item) => {
      try {
        const result = await processor(item);
        onItemSuccess?.(item, result);
        return { success: true, item, result };
      } catch (error) {
        onItemFailure?.(item, error);
        return { success: false, item, error };
      }
    });

    const batchResults = await Promise.all(batchPromises);

    batchResults.forEach((result) => {
      if (result.success) {
        successCount++;
        successfulItems.push(result.result);
      } else {
        failureCount++;
        failedItems.push({ item: result.item, error: result.error });
      }
    });

    onProgress?.(i + batch.length, items.length);
  }

  return { successCount, failureCount, successfulItems, failedItems };
}

export async function processImportBatch(
  items: ImportBatchItem[],
  options: BatchProcessorOptions = {}
): Promise<{
  successCount: number;
  failureCount: number;
  successfulImports: string[];
  failedItems: { item: ImportBatchItem; error: any }[];
}> {
  const successfulImports: string[] = [];

  const result = await processBatch(
    items,
    async (item: ImportBatchItem) => {
      await importDocument(
        item.docInfo,
        item.locale,
        item.data,
        item.translationContext,
        false
      );
      return item.key;
    },
    {
      ...options,
      onItemSuccess: (item: ImportBatchItem, key: string) => {
        successfulImports.push(key);
        options.onItemSuccess?.(item, key);
      },
      onItemFailure: (item: ImportBatchItem, error: any) => {
        console.error(
          `Failed to import ${item.docInfo.documentId} (${item.locale}):`,
          error
        );
        options.onItemFailure?.(item, error);
      },
    }
  );

  return {
    ...result,
    successfulImports,
  };
}
