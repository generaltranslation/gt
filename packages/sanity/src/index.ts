import TranslationsTab from './components/TranslationsTab';
import {
  Secrets,
  Adapter,
  ExportForTranslation,
  ImportTranslation,
  TranslationFunctionContext,
  TranslationsTabConfigOptions,
} from './types';
import { baseDocumentLevelConfig } from './configuration/baseDocumentLevelConfig';
import { baseFieldLevelConfig } from './configuration/baseFieldLevelConfig';
import { findLatestDraft } from './configuration/utils/findLatestDraft';
import { documentLevelPatch } from './configuration/baseDocumentLevelConfig/documentLevelPatch';
import { fieldLevelPatch } from './configuration/baseFieldLevelConfig';
import {
  BaseDocumentSerializer,
  BaseDocumentDeserializer,
  BaseDocumentMerger,
  defaultStopTypes,
  customSerializers,
  SerializedDocument,
} from './serialization';

export type {
  Secrets,
  Adapter,
  ExportForTranslation,
  ImportTranslation,
  TranslationFunctionContext,
  TranslationsTabConfigOptions,
  SerializedDocument,
};
export {
  TranslationsTab,
  //helpers for end developers who may need to customize serialization
  findLatestDraft,
  documentLevelPatch,
  fieldLevelPatch,
  BaseDocumentSerializer,
  BaseDocumentDeserializer,
  BaseDocumentMerger,
  defaultStopTypes,
  customSerializers,
  attachGTData,
  detachGTData,
};

import { GTAdapter } from './adapter';
import { definePlugin } from 'sanity';
import { route } from 'sanity/router';
import { gt, pluginConfig } from './adapter/core';
import { libraryDefaultLocale } from 'generaltranslation/internal';
import type { IgnoreFields, TranslateDocumentFilter } from './adapter/types';
import TranslationsTool from './components/page/TranslationsTool';
import { SECRETS_NAMESPACE } from './utils/shared';
import type { PortableTextHtmlComponents } from '@portabletext/to-html';
import { attachGTData, detachGTData } from './serialization/data';

interface ConfigOptions {
  adapter: Adapter;
  secretsNamespace: string | null;
  exportForTranslation: ExportForTranslation;
  importTranslation: ImportTranslation;
}

export const defaultDocumentLevelConfig: ConfigOptions = {
  ...baseDocumentLevelConfig,
  adapter: GTAdapter,
  secretsNamespace: 'generaltranslation',
};

export const defaultFieldLevelConfig: ConfigOptions = {
  ...baseFieldLevelConfig,
  adapter: GTAdapter,
  secretsNamespace: 'generaltranslation',
};

export { GTAdapter };

export type GTPluginConfig = Omit<
  Parameters<typeof gt.setConfig>[0],
  'locales'
> & {
  locales: string[];
  singletons?: string[];
  // Optional mapping function to map source document ids to translated singleton document ids
  // By default, the translated singleton document is is `${sourceDocumentId}-${locale}`
  singletonMapping?: (sourceDocumentId: string, locale: string) => string;
  ignoreFields?: IgnoreFields[];
  languageField?: string;
  translateDocuments?: TranslateDocumentFilter[];
  secretsNamespace?: string;
  additionalStopTypes?: string[];
  additionalSerializers?: Partial<PortableTextHtmlComponents>;
  additionalDeserializers?: Record<string, any>;
  additionalBlockDeserializers?: any[];
};

/**
 * Usage in `sanity.config.ts` (or .js)
 *
 * ```ts
 * import {defineConfig} from 'sanity'
 * import {gtPlugin} from 'gt-sanity'
 *
 * export default defineConfig({
 *   // ...
 *   plugins: [gtPlugin()],
 * })
 * ```
 */
export const gtPlugin = definePlugin<GTPluginConfig>(
  ({
    languageField = 'language',
    sourceLocale = libraryDefaultLocale,
    locales,
    customMapping,
    apiKey,
    projectId,
    singletons,
    singletonMapping,
    ignoreFields,
    translateDocuments,
    secretsNamespace = SECRETS_NAMESPACE,
    additionalStopTypes = [],
    additionalSerializers = {},
    additionalDeserializers = {},
    additionalBlockDeserializers = [],
  }) => {
    // Validate the translateDocuments
    translateDocuments = translateDocuments?.filter((filter) => {
      if (filter.documentId || filter.type) {
        return true;
      }
      return false;
    });

    pluginConfig.init(
      secretsNamespace,
      languageField,
      sourceLocale,
      locales,
      singletons || [],
      // singletons is a string array of singleton document ids
      singletonMapping ||
        ((sourceDocumentId, locale) => `${sourceDocumentId}-${locale}`),
      ignoreFields || [],
      translateDocuments || [],
      additionalStopTypes,
      additionalSerializers,
      additionalDeserializers,
      additionalBlockDeserializers
    );
    gt.setConfig({
      sourceLocale: sourceLocale,
      customMapping: customMapping,
      apiKey: apiKey,
      projectId: projectId,
    });
    return {
      name: 'gt-sanity',
      tools: [
        {
          name: 'translations',
          title: 'Translations',
          component: TranslationsTool,
          router: route.create('/*'),
        },
      ],
    };
  }
);
