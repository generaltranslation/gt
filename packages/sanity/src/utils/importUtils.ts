import { SanityDocument } from 'sanity';
import { GTFile, Secrets, TranslationFunctionContext } from '../types';
import {
  downloadTranslations,
  BatchedFiles,
} from '../translation/downloadTranslations';
import { processImportBatch, ImportBatchItem } from './batchProcessor';

export interface TranslationStatus {
  progress: number;
  isReady: boolean;
  translationId?: string;
}

export interface ImportResult {
  successCount: number;
  failureCount: number;
  successfulImports: string[];
}

export interface ImportOptions {
  filterReadyFiles?: (key: string, status: TranslationStatus) => boolean;
  onProgress?: (current: number, total: number) => void;
  onImportSuccess?: (key: string) => void;
}

export async function getReadyFilesForImport(
  documents: SanityDocument[],
  translationStatuses: Map<string, TranslationStatus>,
  options: ImportOptions = {}
): Promise<BatchedFiles> {
  const { filterReadyFiles = () => true } = options;
  const readyFiles: BatchedFiles = [];

  for (const [key, status] of translationStatuses.entries()) {
    if (
      status.isReady &&
      status.translationId &&
      filterReadyFiles(key, status)
    ) {
      const [documentId, locale] = key.split(':');
      const document = documents.find(
        (doc) => (doc._id?.replace('drafts.', '') || doc._id) === documentId
      );

      if (document) {
        readyFiles.push({
          documentId,
          versionId: document._rev,
          translationId: status.translationId,
          locale,
        });
      }
    }
  }

  return readyFiles;
}

export async function importTranslations(
  readyFiles: BatchedFiles,
  secrets: Secrets,
  translationContext: TranslationFunctionContext,
  options: ImportOptions = {}
): Promise<ImportResult> {
  if (readyFiles.length === 0) {
    return { successCount: 0, failureCount: 0, successfulImports: [] };
  }

  const downloadedFiles = await downloadTranslations(readyFiles, secrets);

  const importItems: ImportBatchItem[] = downloadedFiles.map((file) => ({
    docInfo: {
      documentId: file.docData.documentId,
      versionId: file.docData.versionId,
    },
    locale: file.docData.locale,
    data: file.data,
    translationContext,
    key: `${file.docData.documentId}:${file.docData.locale}`,
  }));

  const result = await processImportBatch(importItems, {
    onProgress: options.onProgress,
    onItemSuccess: (item, key) => {
      options.onImportSuccess?.(key);
    },
  });

  return {
    successCount: result.successCount,
    failureCount: result.failureCount,
    successfulImports: result.successfulImports,
  };
}
