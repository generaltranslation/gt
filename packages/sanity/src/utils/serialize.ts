import { SanityDocument, Schema } from 'sanity';
import {
  BaseDocumentDeserializer,
  BaseDocumentSerializer,
  defaultStopTypes,
  customSerializers,
  customBlockDeserializers,
} from '../serialization/';

export function deserializeDocument(document: string) {
  const deserialized = BaseDocumentDeserializer.deserializeDocument(
    document,
    { types: {} },
    customBlockDeserializers
  ) as SanityDocument;
  return deserialized;
}

export function serializeDocument(
  document: SanityDocument,
  schema: Schema,
  baseLanguage: string
) {
  const serialized = BaseDocumentSerializer(schema).serializeDocument(
    document,
    'document',
    baseLanguage,
    defaultStopTypes,
    customSerializers
  );
  return serialized;
}
