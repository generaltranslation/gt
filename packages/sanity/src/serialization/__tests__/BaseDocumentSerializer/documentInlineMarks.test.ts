import { PortableTextBlock, SanityDocument } from 'sanity';
import { describe, expect, test } from 'vitest';
import {
  getSerialized,
  getValidFields,
  toPlainText,
  getDeserialized,
} from '../helpers';
import { docWithInlineMarks, findByClass, getHTMLNode, schema } from './utils';
import { attachGTData, BaseDocumentSerializer, customSerializers } from '../..';
import { TranslationLevel } from '../../types';
import { defaultStopTypes } from '../..';
import merge from 'lodash.merge';
import { PortableTextHtmlComponents } from '@portabletext/to-html';

function serialize(document: SanityDocument, level: TranslationLevel) {
  const serializer = BaseDocumentSerializer(schema);
  return serializer.serializeDocument(
    document,
    level,
    'en',
    defaultStopTypes,
    merge(customSerializers, {
      marks: {
        link: ({ value, children }) =>
          attachGTData(`<a>${children}</a>`, value, 'markDef'),
        inlineMath: ({ value, children }) =>
          attachGTData(`<span>${children}</span>`, value, 'markDef'),
      },
    } satisfies Partial<PortableTextHtmlComponents>)
  );
}

const serialized = serialize(docWithInlineMarks, 'document');
const deserialized = getDeserialized(docWithInlineMarks, 'document');
const docTree = getHTMLNode(serialized).body.children[0];

test('Global test of working doc-level functionality and snapshot match', () => {
  expect(serialized).toMatchSnapshot();
});

test('Test of working deserialization and snapshot match', () => {
  expect(deserialized).toMatchSnapshot();
});
/*
 * Top-level plain text
 */
test('String and text types get serialized correctly at top-level', () => {
  const HTMLString = findByClass(docTree.children, 'title');
  const HTMLText = findByClass(docTree.children, 'snippet');
  expect(HTMLString?.innerHTML).toEqual(docWithInlineMarks.title);
  expect(HTMLText?.innerHTML).toEqual(docWithInlineMarks.snippet);
});
