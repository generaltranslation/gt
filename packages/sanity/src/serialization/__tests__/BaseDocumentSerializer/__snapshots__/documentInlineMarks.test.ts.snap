// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Global test of working doc-level functionality and snapshot match 1`] = `
{
  "content": "<html><head><meta name="_id" content="drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a"><meta name="_type" content="documentLevelArticle"><meta name="_rev" content="fxcnqj-uew-gr7-v6r-z965tfm69"><meta name="version" content="3"></head><body><div class="documentLevelArticle" id="drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a" data-type="object"><span class="title">My Document-Level Article</span><span class="snippet">This is text in my text field.</span><div class="tags" data-type="array"><span>tag 1</span><span>tag 2</span><span>tag 3</span></div><div class="content" data-type="array"><p id="a4d177e92666">This is block text with <span class="unknown__pt__mark__linkField">a website link</span> and <span class="unknown__pt__mark__linkField">a post reference</span><span class="unknown__pt__mark__linkField"> at the top level.</span></p></div></div></body></html>",
  "name": "drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a",
}
`;

exports[`Test of working deserialization and snapshot match 1`] = `
{
  "_id": "drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a",
  "_rev": "fxcnqj-uew-gr7-v6r-z965tfm69",
  "_type": "documentLevelArticle",
  "content": [
    {
      "_key": "a4d177e92666",
      "_type": "block",
      "children": [
        {
          "_key": "randomKey-10",
          "_type": "span",
          "marks": [],
          "text": "This is block text with a website link and a post reference at the top level.",
        },
      ],
      "markDefs": [],
      "style": "normal",
    },
  ],
  "snippet": "This is text in my text field.",
  "tags": [
    "tag 1",
    "tag 2",
    "tag 3",
  ],
  "title": "My Document-Level Article",
}
`;
